pkg/blobstore/spanner_blob_access.go

```diff
@@ -0,0 +1,975 @@
+package blobstore
+
+//
+// BuildBarn blob access layer that stores metadata and small (<10MB) blobs in Spanner, while
+// storing large blobs in a GCS bucket.  We rely on Google to delete stale objects, but we avoid
+// the situation where GCS blobs are deleted before the Spanner records referencing them.  We
+// Update reference timestamps at insert time (in Put) and in FindMissing (to ensure blobs live
+// long enough for a bazel build or test to complete).  We also prevent rows scheduled for
+// deletion in GCS from being returned by Get.  Thus, if a blob is stale in GCS, it's metadata
+// should also be stale in Spanner, as long as the GCS TTL is at least as long as the Spanner TTL.
+// To make sure action cache entries are retained on an LRU-like basis, we queue up a list of
+// hashes as they are referenced and periodically update ther reference time.  Doing this one at
+// a time incurs too much overhead in spanner, so we perform bulk operations.  The downside of
+// this approach is that we can lose reference updates if the servers reboot while updates are
+// still queued.  Worst case, these objects will be evicted and need to be rebuilt the next time
+// they are needed.
```

Comment:
I don't think this tradeoff is acceptable.

Reply:
Performance in unacceptable without this (databases make poor file systems).  Potentially rebuilding stuff that is evicted earlier than it might be is an acceptable tradeoff, because it still results in correct operation.  However, I'm open to ideas on how we can mitigate this.